lib/emulator.cpp

```diff
@@ -227,49 +242,42 @@ void emulator::emulator_t::check_and_handle_interrupts() {
-    
-    if (in_m_mode) {
-        if (!mstatus.bit(3)) {//mstatus.MIE is not set, so since we're in M-mode, this means interrupts are disabled
-            irvelog(0, "Interrupts are currently disabled in M-mode");
-            return;
-        }
-    } else if (in_s_mode) {
-        if (!mstatus.bit(1)) {//mstatus.SIE is not set, so since we're in S-mode, this means interrupts are disabled
-            irvelog(0, "Interrupts are currently disabled in S-mode");
-            return;
-        }
-    }
-    //If we reach this point, interrupts aren't globally disabled, so we need to check for them
+    auto interrupt_regs = this->m_CSR.fast_implicit_read_interrupt_regs();
 
-    //
-    //Check if any interrupts are "interrupting", and choose the one with the highest priority
-    //
+    Reg mstatus = interrupt_regs.mstatus;
+    bool in_m_mode = interrupt_regs.privilege_mode == PrivilegeMode::MACHINE_MODE;
+    bool in_s_mode = interrupt_regs.privilege_mode == PrivilegeMode::SUPERVISOR_MODE;
+
+    bool global_enable_for_current_privilege = (in_m_mode && (mstatus.bit(3) == 1)) || (in_s_mode && (mstatus.bit(1) == 1));
 
     //Get mip and sie. NOTE: We don't need to read sip and sie, since those are just shadows for S-mode code to use
-    Reg mip = this->m_CSR.implicit_read(Csr::Address::MIP);
-    Reg mie = this->m_CSR.implicit_read(Csr::Address::MIE);
+    Reg mip = interrupt_regs.mip;
+    Reg mie = interrupt_regs.mie;
 
     //Also mideleg will be useful
-    Reg mideleg = this->m_CSR.implicit_read(Csr::Address::MIDELEG);
+    Reg mideleg = interrupt_regs.mideleg;
 
     //Helper lambda. Returns true if the given bit is "interrupting". Aka, that...
     //1. The interrupt is pending (mip/sip)
     //2. The interrupt is enabled (mie/sie)
-    //3. The interrupt is delegated to the current or higher privilege level (mideleg/sideleg)
+    //3. Interrupts are globally enabled at this level, or the interrupt is for a higher privilege level (mstatus.MIE/SIE)
+    //4. The interrupt is delegated to the current or a higher privilege level (mideleg/sideleg)
     //Also sets `delegated_to_smode` properly for later use
     auto is_interrupting = [&](uint8_t bit) {
         assert((bit < 32) && "Invalid is_interrupting() bit!");
-        bool pending = mip.bit(bit) == 1;
-        bool enabled = mie.bit(bit) == 1;
-
-        //The only time we'd ignore an interrupt is if we're in M-mode and it's delegated to S-mode
+        bool pending            = mip.bit(bit) == 1;
+        bool enabled            = mie.bit(bit) == 1;
         bool delegated_to_smode = mideleg.bit(bit) == 1;
-        bool ignored_in_current_mode = (in_m_mode && delegated_to_smode);
 
-        bool interrupting = pending && enabled && !ignored_in_current_mode;
+        //Either in user mode, or in s-mode but it is delegated to m-mode
+        bool interrupt_is_for_higher_privilege_level = !in_m_mode && (!in_s_mode || (in_s_mode && !delegated_to_smode));
```

Comment:
Logic would be much clearer here if we also had `in_u_mode`

Reply:
Good point, but I don't want to introduce a bug if I change it now, and I don't think it's that unreadable.

Basically I read it as "An interrupt can't be for a higher privilege level if we're already in m-mode. If we aren't, then the only way it's for a higher privilege level is if we're in u-mode or we're in s-mode but it's delegated to m-mode."